# MỤC TIÊU

Thực hiện bài toán kiểm định 1 phía cho 2 tham số

- Kỳ vọng ($\mu$)
- Tỷ lệ (p)

# VÍ DỤ

## Ví dụ 1

![image.png](attachment:a05ecccb-fe85-42f9-9ddb-d2e30684cefb.png)

In [6]:
import numpy as np
import scipy.stats as st

In [7]:
freq_table = [(155, 3), (165, 9), (175, 11), (185, 3), (195, 2), (205, 1)]
# freq_table = [(155, 3), (165, 2), (175, 1)]

In [9]:
data = []
for v, f in freq_table:
    for i in range(f):
        data.append(v)

# data

In [10]:
# Tinh toan cac so lieu thong ke tren mau

# Kich thuoc mau
n = len(data)

# Tinh trung binh mau
tb = round(np.mean(data), 4)

# Tinh do lech chuan mau
dlc = round(np.std(data, ddof=1), 4)

print('Kich thuoc mau: ', n)
print('Trung binh mau: ', tb)
print('Do lech chuan mau: ', dlc)

Kich thuoc mau:  29
Trung binh mau:  173.2759
Do lech chuan mau:  11.9729


In [ ]:
import math

In [ ]:
# Cau b: Tim khoang uoc luong cho ky vong o do tin cay 95%
st.t.interval(0.95, df=n-1, loc=tb, scale=(dlc/math.sqrt(n)))

(168.72165240566306, 177.83014759433695)

In [ ]:
# Cau b (tinh tay)
# Vì ta không biết giá trị của sigma (độ lệch chuẩn quần thể),
# vì vậy, trung bình mẫu tuân theo phân phối Student
# Ta có: sai số ước lượng:
epsilon = st.t.isf(0.025, df=n-1) * (dlc / math.sqrt(n))

epsilon

4.554238369648043

In [ ]:
# Khoảng ước lượng:
print('Khoảng ước lượng: ({0}, {1})'.format(tb-epsilon, tb+epsilon))

Khoảng ước lượng: (168.72166163035197, 177.83013836964804)


__Câu c: Kiểm định__

Phát biểu: "Lượng cholesterol trung bình là 175mg%".

B1: Đặt cập giả thuyết, đối thuyết

Ho = $\mu = 175$

Ha = $\mu <> 175$

Giả định Ho đúng, tìm bằng chứng để thử bác bỏ Ho

B2: Mức ý nghĩa của kiểm định: $\alpha$ = 0.05

B4: Xác định miền bác bỏ giả thuyết:        

In [ ]:
t_alpha_chia_2 = st.t.isf(0.025, df=n-1)
t_alpha_chia_2

2.048407141795244

Vậy miền bác bỏ của giả thuyết là: {t: |t| > 2.0484}}

In [ ]:
# B3: Tính trị kiểm định:

t = (tb - 175) / (dlc / math.sqrt(n))
t

-0.7754648116981316

B5: Kết luận

t không thuộc miền bác bỏ.

Vậy chấp nhận giả thuyết Ho ban đầu

## Ví dụ 2

# BÀI TẬP

## Kiểm định đặc trưng tỉ lệ (p)

### Bài 1

Dataset 18 (__M&M__) chứa thông tin trọng lượng và màu sắc của mẫu các viên kẹo của hãng M&M.

Sử dụng mẫu trên bạn hãy kiểm tra phát biểu sau: __"20% kẹo M&M có màu đỏ"__ với mức ý nghĩa là 5%

*Hướng dẫn: tính tỷ lệ các viên kẹo có màu đỏ từ dữ liệu, sử dụng kết quả để kiểm định phát biểu trên*

In [20]:
import numpy as np
from scipy import stats

# Dữ liệu màu sắc của các viên kẹo M&M
red_data = [0.751, 0.841, 0.856, 0.799, 0.966, 0.859, 0.857, 0.942, 0.873, 0.809, 0.890, 0.878, 0.905]

# Tỷ lệ trung bình của màu đỏ
red_mean = np.mean(red_data)

# Giả thuyết không: Tỷ lệ các viên kẹo có màu đỏ là 20%
null_hypothesis = 0.20

# Độ lệch chuẩn
std_dev = np.std(red_data, ddof=1)

# Kích thước mẫu
sample_size = len(red_data)

# Tính giá trị z
z = (red_mean - null_hypothesis) / (std_dev / np.sqrt(sample_size))

# Tính p-value
p_value = 1 - stats.norm.cdf(z)

# Mức ý nghĩa
alpha = 0.05

# Kiểm tra mức ý nghĩa
if p_value < alpha:
    print(f"Phát biểu '20% kẹo M&M có màu đỏ' không đúng với mức ý nghĩa {alpha}")
else:
    print(f"Phát biểu '20% kẹo M&M có màu đỏ' không bị bác bỏ với mức ý nghĩa {alpha}")









Phát biểu '20% kẹo M&M có màu đỏ' không đúng với mức ý nghĩa 0.05


### Bài 2

Dataset 3 (__FRESH15__) chứa thông tin khảo sát những thay đổi về trọng lượng (weight - WT) và chỉ số khối của cơ thể (BMI) của sinh viên năm nhất tại một trường Đại Học. Thời điểm khảo sát là bắt đầu học kỳ (tháng 9 - SEP) và kết thúc học kỳ (tháng 4 - APR), các sinh viên được chọn ngẫu nhiên trong trường.

Sử dụng mẫu trên, hãy kiểm tra phát biểu:__"50% số sinh viên trong trường là nam"__ với mức ý nghĩa 50%

Các thuộc tính trong bộ dữ liệu:
- SEX: Giới tính
- WTSEP: trọng lượng đo vào tháng 9 (mới nhập học)
- WTAPR: trọng lượng đo vào tháng 4 (kết thúc năm học)
- BMISP: BMI tháng 9
- BMIAP: BMI tháng 4

In [25]:
import pandas as pd
from scipy import stats

# Tạo DataFrame từ dữ liệu
pf = pd.read_excel('/content/03_FRESH15.xls')

# Số lượng nam
num_males = len(pf[pf['SEX'] == 'M'])

# Tổng số sinh viên
total_students = len(pf)

# Tỷ lệ nam
male_ratio = num_males / total_students

# Giả thuyết không: Tỷ lệ nam là 50%
null_hypothesis = 0.50

# Độ lệch chuẩn ước tính cho tỷ lệ nam
std_error = np.sqrt(male_ratio * (1 - male_ratio) / total_students)

# Tính giá trị z
z = (male_ratio - null_hypothesis) / std_error

# Tính p-value
p_value = 2 * (1 - stats.norm.cdf(abs(z)))  # Nhân với 2 vì đây là kiểm định 2 đuôi

# Mức ý nghĩa
alpha = 0.05

# Kiểm tra mức ý nghĩa
if p_value < alpha:
    print(f"Phát biểu '50% số sinh viên nam' không đúng với mức ý nghĩa {alpha}")
else:
    print(f"Phát biểu '50% số sinh viên nam' không bị bác bỏ với mức ý nghĩa {alpha}")




Phát biểu '50% số sinh viên nam' không bị bác bỏ với mức ý nghĩa 0.05


### Bài 3

Dataset 6(__BEARS__) chứa thông tin khảo sát ngẫu nhiên các con gấu trong một khu rừng.

Sử dụng mẫu trên, bạn hãy kiểm tra phát biểu: __"tỷ lệ gấu đực trong khu rừng là 50%""__ với mức ý nghĩa 90%.

Biết rằng thuộc tính SEX có hai giá trị: 1 - gấu đực, 2- gấu cái.

### Bài 4

Dataset 9 __(MOVIES)__ chứa thông tin khảo sát một số phim trong khoảng thời gian 33 năm gần đây.

Có ý kiến cho rằng: __"55% các phim trong khoảng thời gian 33 năm gần đây có nhãn R (là phim giới hạn độ tuổi - Restricted)"__ .Theo bạn với mức ý nghĩa 95% ý kiến trên có đúng không?

Tham khảo: [Hệ thống phân loại của hiệp hội điện ảnh Mỹ](https://vi.wikipedia.org/wiki/H%E1%BB%87_th%E1%BB%91ng_ph%C3%A2n_lo%E1%BA%A1i_phim_c%E1%BB%A7a_MPAA#Ph%C3%A2n_lo%E1%BA%A1i_v%E1%BB%9Bi_c%C3%A1c_trailer_phim)

## Kiểm định đặc trưng kỳ vọng ($\mu$) biết ($\sigma$)

### Bài 5

Dataset 13 __(VOLTAGE)__ liệt kê các mức điện áp được đo tại các hộ gia đình bởi các máy đo khác nhau: tại nhà (HOME), tại máy phát độc lập (Generator), tại trạm phát (UPS).

Công ty cung cấp điện cho vùng này cam kết rằng: __"Mục tiêu cung cấp điện cho mỗi hộ gia đình là 120V"__ . Biết rằng độ lệch chuẩn của tất cả các mức điện áp đo được là 0.24V.

Với mức ý nghĩa 1%. Bạn hãy kiểm tra xem công ty có làm đúng cam kết hay không?

Hướng dẫn: Tính múc điện áp trung bình tại nhà (HOME), và sử dụng để kiểm định cam kết

In [27]:
import numpy as np
from scipy import stats

# Dữ liệu mức điện áp tại nhà (HOME)
voltage_data_home = [
    123.8, 123.9, 123.9, 123.3, 123.4, 123.3, 123.3, 123.6, 123.5, 123.5,
    123.5, 123.7, 123.6, 123.7, 123.9, 124.0, 124.2, 123.9, 123.8, 123.8,
    124.0, 123.9, 123.6, 123.5, 123.4, 123.4, 123.4, 123.4, 123.3, 123.3,
    123.5, 123.6, 123.8, 123.9, 123.9, 123.9, 123.8, 123.9, 123.7, 123.8,
    123.8, 123.8
]

# Cam kết của công ty
target_voltage = 120  # 120V

# Độ lệch chuẩn của mức điện áp đo được
std_dev = 0.24

# Tính mức điện áp trung bình tại nhà
mean_voltage_home = np.mean(voltage_data_home)

# Số lượng mẫu
n = len(voltage_data_home)

# Tính giá trị z-score
z_score = (mean_voltage_home - target_voltage) / (std_dev / np.sqrt(n))

# Tính p-value
p_value = 2 * (1 - stats.norm.cdf(abs(z_score)))  # Nhân với 2 vì đây là kiểm định 2 đuôi

# Mức ý nghĩa
alpha = 0.01  # 1%

# Kiểm tra mức ý nghĩa
if p_value < alpha:
    print(f"Công ty không làm đúng cam kết với mức ý nghĩa {alpha}")
else:
    print(f"Công ty làm đúng cam kết với mức ý nghĩa {alpha}")



Công ty không làm đúng cam kết với mức ý nghĩa 0.01


## Kiểm định đặc trưng kỳ vọng ($\mu$) không biết ($\sigma$)

### Bài 6

Dataset 13 __(VOLTAGE)__ liệt kê các mức điện áp được đo tại các hộ gia đình bởi các máy đo khác nhau: tại nhà (HOME), tại máy phát độc lập (Generator), tại trạm phát (UPS).

Công ty cung cấp điện cho vùng này cam kết rằng: __"Mục tiêu cung cấp điện cho mỗi hộ gia đình là 120V"__ .

Với mức ý nghĩa 1%. Bạn hãy kiểm tra xem công ty có làm đúng cam kết hay không?

In [32]:
import numpy as np
from scipy import stats

# Dữ liệu mức điện áp tại các hộ gia đình (các cột: HOME, Generator, UPS)
voltage_data = [
    [123.8, 124.8, 123.1],
    [123.9, 124.3, 123.1],
    [123.9, 125.2, 123.6],
    [123.3, 124.5, 123.6],
    [123.4, 125.1, 123.6],
    [123.3, 124.8, 123.7],
    [123.3, 125.1, 123.7],
    [123.6, 125.0, 123.6],
    [123.5, 124.8, 123.6],
    [123.5, 124.7, 123.8],
    [123.5, 124.5, 123.7],
    [123.7, 125.2, 123.8],
    [123.6, 124.4, 123.5],
    [123.7, 124.7, 123.7],
    [123.9, 124.9, 123.0],
    [124.0, 124.5, 123.8],
    [124.2, 124.8, 123.8],
    [123.9, 124.8, 123.1],
    [123.8, 124.5, 123.7],
    [123.8, 124.6, 123.7],
    [124.0, 125.0, 123.8],
    [123.9, 124.7, 123.8],
    [123.6, 124.9, 123.7],
    [123.5, 124.9, 123.8],
    [123.4, 124.7, 123.7],
    [123.4, 124.2, 123.8],
    [123.4, 124.7, 123.8],
    [123.4, 124.8, 123.8],
    [123.3, 124.4, 123.9],
    [123.3, 124.6, 123.8],
    [123.5, 124.4, 123.9],
    [123.6, 124.0, 123.9],
    [123.8, 124.7, 123.9],
    [123.9, 124.4, 123.9],
    [123.9, 124.6, 123.6],
    [123.8, 124.6, 123.2],
    [123.9, 124.6, 123.1],
    [123.7, 124.8, 123.0],
    [123.8, 124.3, 122.9],
    [123.8, 124.0, 123.0]
]

# Cam kết của công ty
target_voltage = 120  # 120V

# Số lượng mẫu
n = len(voltage_data)

# Tính mức điện áp trung bình tại tất cả các hộ gia đình
mean_voltage_all = np.mean(voltage_data)

# Tính độ lệch chuẩn của mức điện áp đo được
std_dev = np.std(voltage_data, ddof=1)  # ddof=1 cho phép tính độ lệch chuẩn mẫu

# Tính giá trị t-statistic
t_statistic = (mean_voltage_all - target_voltage) / (std_dev / np.sqrt(n))

# Tính p-value
p_value = 2 * (1 - stats.t.cdf(abs(t_statistic), df=n - 1))  # Nhân với 2 vì đây là kiểm định 2 đuôi

# Mức ý nghĩa
alpha = 0.01  # 1%

# Kiểm tra mức ý nghĩa
if p_value < alpha:
    print(f"Công ty không làm đúng cam kết với mức ý nghĩa {alpha}")
else:
    print(f"Công ty làm đúng cam kết với mức ý nghĩa {alpha}")



Công ty không làm đúng cam kết với mức ý nghĩa 0.01


### Bài 7

Dataset 2 __(BODYTEMP)__ chứa thông tin khảo sát nhiệt độ cơ thể từ một nhóm người vào cùng một thời điểm lúc 12h trưa.

Có một phát biểu nói rằng, nhiệt độ cơ thể người đo lúc 12h trưa trong quần thể này là 98,6°

Dựa vào mẫu trên, bạn hãy cho biết phát biểu trên đúng hay sai ở mức ý nghĩa là 5%.

In [33]:
import numpy as np
from scipy import stats

# Dữ liệu nhiệt độ cơ thể
body_temp_data = [
    98.6, 98.6, 98.0, 98.0, 99.0, 98.4, 98.4, 98.4, 98.4, 98.6,
    98.6, 98.8, 98.6, 97.0, 97.0, 98.8, 97.6, 97.7, 98.8, 98.0,
    98.0, 98.3, 98.5, 97.3, 98.7, 97.4, 98.9, 98.6, 99.5, 97.5,
    97.3, 97.6, 98.2, 99.6, 98.7, 99.4, 98.2, 98.0, 98.6, 98.6,
    97.2, 98.4, 98.6, 98.2, 98.0, 97.8, 98.0, 98.4, 98.6, 98.6,
    97.8, 99.0, 96.5, 97.6, 98.0, 96.9, 97.6, 97.1, 97.9, 98.4,
    97.3, 98.0, 97.5, 97.6, 98.2, 98.5, 98.8, 98.7, 97.8, 98.0,
    97.1, 97.4, 99.4, 98.4, 98.6, 98.4, 98.5, 98.6, 98.3, 98.7,
    98.8, 99.1, 98.6, 97.9, 98.8, 98.0, 98.7, 98.5, 98.9, 98.4,
    98.6, 97.1, 97.9, 98.8, 98.7, 97.6, 98.2, 99.2, 97.8, 98.0,
    98.4, 97.8, 98.4, 97.4, 98.0, 97.0
]

# Giá trị tham chiếu
reference_temp = 98.6  # Giá trị tham chiếu là 98.6°F

# Tính kiểm định t-test một mẫu
t_statistic, p_value = stats.ttest_1samp(body_temp_data, reference_temp)

# Mức ý nghĩa
alpha = 0.05  # 5%

# Kiểm tra mức ý nghĩa
if p_value < alpha:
    print(f"Phát biểu sai ở mức ý nghĩa {alpha}")
else:
    print(f"Phát biểu đúng ở mức ý nghĩa {alpha}")


Phát biểu sai ở mức ý nghĩa 0.05


### Bài 8

__FICO Score__ là đại lượng dùng để đo khả năng tín dụng của một cá nhân hoặc một tập thể. Sử dụng FICO Score, ngân hàng có thể quyết định xem có nên cho khách hàng mượn tiền hay không?

Dataset 24 __(FICO)__ chứa thông tin điểm tín dụng (FICO Score) của một số cá nhân trong vùng.

Sử dụng dữ liệu mẫu từ, bạn hãy cho biết ý kiến: __"điểm tín dùng trung bình của quần thể là 678"__ là đúng hay sai với mức ý nghĩa 5%

In [34]:
import numpy as np
from scipy import stats

# Dữ liệu FICO Score
fico_scores = [
    708, 713, 781, 809, 797, 793, 711, 681, 768, 611,
    698, 836, 768, 532, 657, 559, 741, 792, 701, 753,
    745, 681, 598, 693, 743, 444, 502, 739, 755, 835,
    714, 517, 787, 714, 497, 636, 637, 797, 568, 714,
    618, 830, 579, 818, 654, 617, 849, 798, 751, 731,
    850, 591, 802, 756, 689, 789, 628, 692, 779, 756,
    782, 760, 503, 784, 591, 834, 694, 795, 660, 651,
    696, 638, 635, 795, 519, 682, 824, 603, 709, 777,
    829, 744, 752, 783, 630, 753, 661, 604, 729, 722,
    706, 594, 664, 782, 579, 796, 611, 709, 697, 732
]

# Giá trị tham chiếu
reference_value = 678  # Giá trị tham chiếu là 678

# Tính kiểm định t-test một mẫu
t_statistic, p_value = stats.ttest_1samp(fico_scores, reference_value)

# Mức ý nghĩa
alpha = 0.05  # 5%

# Kiểm tra mức ý nghĩa
if p_value < alpha:
    print(f"Phát biểu sai ở mức ý nghĩa {alpha}")
else:
    print(f"Phát biểu đúng ở mức ý nghĩa {alpha}")


Phát biểu sai ở mức ý nghĩa 0.05


---